### DISCUSSION TITLE
[SDK] Allow overriding estimateGas Transaction signature

### DISCUSSION THREAD

**Discussion by hbriese** [zkSync-Community-Hub/zksync-developers#81] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Python and Java SDK as they are being updated.

### Issue Description

The JS `Provider.estimateGas` function always estimates a transaction's gas with a 65 byte signature (assumably for EOAs), and disregards any `customData.customSignature`

### Expected Behavior

`customData.customSignature` is respected for `estimateGas` calls

### Code Example

```ts
provider.estimateGas({
      ...
      type: EIP712_TX_TYPE,
      customData: {
        customSignature: '0x...'
      },
    }),
```

**Comment by dutterbutter** [comment]
Hmm, that's interesting 🤔. Thanks for pointing it out! I'll need some time to investigate this behaviour🕵️‍♂️. I'll post back here with any insights as to why that might be or if we _should_ `estimateGas` accordingly. Thank you for the post 💯

**Comment by MexicanAce** [comment, marked as answer]
Hi @hbriese 👋 !

The way that zkSync Era estimates gas is [via a Binary Search algorithm](https://github.com/matter-labs/zksync-era/blob/e575ec101fe88627ab541a52464ab5025c16e6b4/core/bin/zksync_core/src/api_server/tx_sender/mod.rs#L706-L739) that modifies the `gas_limit` of the Transaction ([See here for code](https://github.com/matter-labs/zksync-era/blob/e575ec101fe88627ab541a52464ab5025c16e6b4/core/bin/zksync_core/src/api_server/tx_sender/mod.rs#L550-L560)).

Because of this, the signature will never match as expected, and this also leads to things like [needing to update the nonce of the transaction while it runs inside of a Sandbox](https://github.com/matter-labs/zksync-era/blob/main/core/bin/zksync_core/src/api_server/execution_sandbox/apply.rs#L78-L84). [As you can see in the underlying contract](https://github.com/matter-labs/era-system-contracts/blob/7658f9a18e4642e04c2e06e9468ca111d48ea1f0/bootloader/bootloader.yul#L2279-L2292), the signature is never fully verified as the underlying data (e.g. `gas_limit`) has been modified.

**Comment by hbriese** [comment]
@MexicanAce thanks for your answer.
The problem arrises when additional data is passed in via the signature - not just a single ecdsa secpk1 signature.
In my case this additional data causes extra verification to be performed, thereby affecting the gas usage.

The expected behaviour is for the `customSignature` field to be respected (not replaced with a 65B signature), but for (atleast parts of) the signature passed in to be incorrect as the transaction's `gas_limit` is modified during estimation

**Reply by hbriese** [reply]
@MexicanAce

**Reply by MexicanAce** [reply]
Thank you for the ping (apologies for not responding sooner). Understood, and I'll bring this up internally to better understand the limitations here.

**Comment by dougtende** [comment]
To address the issue with the JS Provider where the estimateGas function does not respect customSignature in customData for transactions, it's crucial to review the implementation of the estimateGas function in the provider codebase. Ensure that the function considers customSignature when estimating gas for transactions, allowing for more accurate gas estimations. This adjustment will align the behavior with the expected functionality, enhancing the overall user experience when interacting with the provider.

**Comment by go4comdathern** [comment]
We acknowledge the issue with the JS Provider's estimateGas function on the Testnet. We are working on resolving it to ensure that customData.customSignature is respected for estimateGas calls. Thank you for your patience and understanding.

**Comment by robbmoser19** [comment]
Acknowledged. The issue with the JS Provider's estimateGas function on Testnet is noted. We're working to ensure that customData.customSignature is properly considered for estimateGas calls. Thank you for your understanding as we address this issue.

**Comment by springsj** [comment]
Ensure that the custom signature is properly included in the transaction data when using the estimateGas function in the JS Provider.

**Comment by inutpediflinda** [comment]
The JS Provider.estimateGas function currently ignores custom signatures in the customData field.

**Comment by rafordtetseyd** [comment]
The issue is related to the JS Provider.estimateGas function on the testnet, which always estimates a transaction's gas with a 65 byte signature for EOAs and disregards any customData.customSignature. The expected behavior is that customData.customSignature should be respected for estimateGas calls. An example code snippet that could be affected is: provider.estimateGas({
  ...
  type: EIP712_TX_TYPE,
  customData: {
    customSignature: '0x...'
  },
}),It is noted that issues are expected with the Python and Java SDK on the testnet as they are being updated. However, no repository link was provided
